In [1]:
#Import libraries
import pandas as pd
import numpy as np

In [8]:
# 1 Create normalization
def normalize(data):
    """
    Normalizes a vector so all values are between 0 and 1.
    """

    data = np.array(data)

    normalized = (data - np.min(data)) / (np.max(data) - np.min(data))

    return normalized

In [9]:
#Test
sample = [5, 10, 15, 20, 25]

print(normalize(sample))

[0.   0.25 0.5  0.75 1.  ]


In [10]:
# 2 Create standardization function
def standardize(data):
    """
    Standardizes a vector using the z-score formula.
    """

    data = np.array(data)

    standardized = (data - np.mean(data)) / np.std(data, ddof=1)

    return standardized

In [11]:
#Test
sample = [5, 10, 15, 20, 25]

print(standardize(sample))

[-1.26491106 -0.63245553  0.          0.63245553  1.26491106]


In [12]:
#Load dataset
housing = pd.read_csv("calif_housing_data.csv")

In [13]:
#Preview
housing.head()

,housing_median_age,total_bedrooms,households,median_income,median_house_value
0,41,129.0,126,8.3252,452600.0
1,21,1106.0,1138,8.3014,358500.0
2,52,190.0,177,7.2574,352100.0
3,52,235.0,219,5.6431,341300.0
4,52,280.0,259,3.8462,342200.0


In [15]:
#3A 
rows = housing.shape[0]

print("Number of rows:", rows)

Number of rows: 20640


In [16]:
#3B
target = housing["median_house_value"]

print(target)

0        452600.0
1        358500.0
2        352100.0
3        341300.0
4        342200.0
           ...   
20635     78100.0
20636     77100.0
20637     92300.0
20638     84700.0
20639     89400.0
Name: median_house_value, Length: 20640, dtype: float64


In [17]:
#3C
housing["bedrooms_per_household"] = (
    housing["total_bedrooms"] /
    housing["households"]
)

housing.head()

,housing_median_age,total_bedrooms,households,median_income,median_house_value,bedrooms_per_household
0,41,129.0,126,8.3252,452600.0,1.023810
1,21,1106.0,1138,8.3014,358500.0,0.971880
2,52,190.0,177,7.2574,352100.0,1.073446
3,52,235.0,219,5.6431,341300.0,1.073059
4,52,280.0,259,3.8462,342200.0,1.081081


#3C This represents the average number of bedrooms per household within each California block group. It provides information about housing density and the size of homes in each neighborhood.

In [18]:
#3D 
housing_features = housing[
    [
        "housing_median_age",
        "median_income",
        "bedrooms_per_household"
    ]
]

housing_features.head()

,housing_median_age,median_income,bedrooms_per_household
0,41,8.3252,1.023810
1,21,8.3014,0.971880
2,52,7.2574,1.073446
3,52,5.6431,1.073059
4,52,3.8462,1.081081


In [19]:
#3E
#Apply standardization to each feature
standardized_df = housing_features.copy()

for column in standardized_df.columns:
    standardized_df[column] = standardize(
        standardized_df[column]
    )

standardized_df.head()

,housing_median_age,median_income,bedrooms_per_household
0,0.982119,2.344709,NaN
1,-0.607004,2.332181,NaN
2,1.856137,1.782656,NaN
3,1.856137,0.932945,NaN
4,1.856137,-0.012881,NaN


In [20]:
#verify means
standardized_df.mean()

housing_median_age        6.609700e-17
median_income             6.058892e-17
bedrooms_per_household             NaN
dtype: float64

In [21]:
#verify std dev
standardized_df.std()

housing_median_age        1.0
median_income             1.0
bedrooms_per_household    NaN
dtype: float64

In [22]:
#display summary
standardized_df.describe()

,housing_median_age,median_income,bedrooms_per_household
count,2.064000e+04,2.064000e+04,0.0
mean,6.609700e-17,6.058892e-17,NaN
std,1.000000e+00,1.000000e+00,NaN
min,-2.196127e+00,-1.774256e+00,NaN
25%,-8.453727e-01,-6.881019e-01,NaN
50%,2.864502e-02,-1.767908e-01,NaN
75%,6.642943e-01,4.592952e-01,NaN
max,1.856137e+00,5.858144e+00,NaN


In [27]:
#a
print("Number of rows:", housing.shape[0])

Number of rows: 20640


In [28]:
#b target vector
#this is the variable the model is predicting
housing["median_house_value"]

0        452600.0
1        358500.0
2        352100.0
3        341300.0
4        342200.0
           ...   
20635     78100.0
20636     77100.0
20637     92300.0
20638     84700.0
20639     89400.0
Name: median_house_value, Length: 20640, dtype: float64

In [33]:
#c
housing["bedrooms_per_household"] = (
    housing["total_bedrooms"] /
    housing["households"]
)

In [34]:
#d
housing_features = housing[
    [
        "housing_median_age",
        "median_income",
        "bedrooms_per_household"
    ]
]

In [35]:
#e
standardized_df = housing_features.copy()

for column in standardized_df.columns:
    standardized_df[column] = standardize(
        standardized_df[column]
    )

standardized_df.head()

,housing_median_age,median_income,bedrooms_per_household
0,0.982119,2.344709,NaN
1,-0.607004,2.332181,NaN
2,1.856137,1.782656,NaN
3,1.856137,0.932945,NaN
4,1.856137,-0.012881,NaN


A - 20640
B - 20637 
C - This feature represents the average number of bedrooms per household in each neighborhood.